### PR TITLE
Do not allow a non-keyed service to be injected to a keyed parameter

### LINKED ISSUE #102204: Different behavior of FromKeyedServices attribute in Minimal APIs handler vs class library
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If we use the `FromKeyedServices` attribute in a class library and the keyed registration is missing, it falls back to the default dependency injection behavior and gets the last registered implementation (even with no key). However, If we use `FromKeyedServices` directly in a Minimal API endpoint handler and the key does not exist, it throws an exception.

### Expected Behavior

The `FromKeyedServices` attribute should behave in the same way in every context.

### Steps To Reproduce

Minimal repro here: https://github.com/marcominerva/KeyedServicesIssue

### Exceptions (if any)

When using `FromKeyedServices` directly in a Minimal API endpoint handler:

System.InvalidOperationException: No service for type 'KeyedServicesIssue.Library.IDependency' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetRequiredKeyedService(Type serviceType, Object serviceKey, ServiceProviderEngineScope serviceProviderEngineScope)
   at lambda_method2(Closure, Object, HttpContext)

When using  `FromKeyedServices` in a class library, no exception at all is thrown, even if the key does not exist.

### .NET Version

8.0.204

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/102204

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/KeyedDependencyInjectionSpecificationTests.cs [hunk 1]
```diff
@@ -263,6 +263,70 @@ public void ResolveKeyedServiceSingletonInstanceWithKeyedParameter()
...
+            serviceCollection.AddSingleton<OtherServiceWithDefaultCtorArgs>();
+
+            var provider = CreateServiceProvider(serviceCollection);
+
+            Assert.Null(provider.GetService<IService>());
+            Assert.NotNull(provider.GetService<OtherServiceWithDefaultCtorArgs>());
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceWithKeyedParameter_MissingRegistrationButWithUnkeyedService()
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/KeyedDependencyInjectionSpecificationTests.cs:314]: The above tests were *not* failing before this PR, but this test and the other new test in "ServiceProviderContainerTests.cs" were.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs [hunk 2]
```diff
@@ -1290,6 +1290,27 @@ public void ScopedServiceResolvedFromSingletonAfterCompilation3()
...
+
+            serviceCollection.AddSingleton<KeyedDependencyInjectionSpecificationTests.OtherService>();
+
+            AggregateException ex = Assert.Throws<AggregateException>(() => serviceCollection.BuildServiceProvider(new ServiceProviderOptions
+            {
+                ValidateOnBuild = true
+            }));
+
+            Assert.Equal(1, ex.InnerExceptions.Count);
+            Assert.Contains("ServiceType: Microsoft.Extensions.DependencyInjection.Specification.KeyedDependencyInjectionSpecificationTests+OtherService", ex.Message);
```
Comment (buyaa-n) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs:1310]: FYI: this test is failing on .NET framework run

Comment (steveharter) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs]: Thanks

**Review by ericstj — APPROVED** [review]
Code change looks good.  Thank you for marking this breaking.  Let's make sure to get the breaking change doc filled along with a workaround.

**Comment by steveharter** [comment]
Breaking change doc at: https://github.com/dotnet/docs/issues/42313

**Pull request merged**